src/result.h

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) 2022 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_RESULT_H
+#define BITCOIN_RESULT_H
+
+#include <optional>
+#include <util/translation.h>
+#include <variant>
+
+/*
+ * 'BResult' is a generic class useful for wrapping a return object
+ * (in case of success) or propagating the error cause.
+*/
+template<class T>
+class BResult {
+private:
+    std::variant<bilingual_str, T> m_variant;
+
+public:
+    BResult() : m_variant(Untranslated("")) {}
+    BResult(const T& _obj) : m_variant(_obj) {}
+    BResult(const bilingual_str& error) : m_variant(error) {}
+
+    /* In case of success, the result object */
+    const T* GetObj() const { return std::get_if<T>(&m_variant); }
+    /* Whether the function succeeded or not */
+    bool GetRes() const { return !std::holds_alternative<bilingual_str>(m_variant); }
+    /* In case of failure, the error cause */
+    bilingual_str GetError() const {
+        auto error = std::get_if<bilingual_str>(&m_variant);
+        return (error ? *error : bilingual_str());
+    }
+
+    explicit operator bool() const { return GetRes(); }
```

Comment:
I wonder if `operator*` should be added, like https://en.cppreference.com/w/cpp/memory/unique_ptr/operator*

but instead of UB, it runs into an assertion failure?

Reply:
Hmm, it could be handy but might end up with users trying to access the object without checking the result existence first (which, as not every error path is covered by tests, could end up badly).

I would rather walk into adding a generic `std::visit` direction, so users are always forced to provide handlers for both outcomes. Which would avoid the pointers need at all and make the code less error prone.